Enable build on Windows x64 (Mingw compiler) - with reviewed changes (#54)
beehive-lab/TornadoVM: assembly/src/docs/20_INSTALL_WINDOWS_WITH_GRAALVM.md

# Build TornadoVM with Graal 20.2.0 JDK 8 / 11 on Windows x64

> ## Install tools

Author: Out of this list all items (except GPU itself) are mentioned in the text as detailed steps how to get and configure them, so this list would only add confusion -- these are not "pre"-requisites according to document.

> ## Install tools

Reviewer: This suggestion is in order to keep this document as standalone and direct Windows users to follow all necessary steps without looking at the files. Let's keep it as it is, and we will rethink in another PR how to structure the documentation to make it cleaner for users. No worry.

> ## Install tools

Reviewer 2: No, it does not add confusion. It is a summary of all dependencies needed to build TornadoVM from the source code. Then, it is ok to explain how to get all the dependencies for each OS/Configuration. I would keep it. Also, we should keep it for consistency.

> ## Install tools

Author: Please put a suggestion and I'll commit it

> ## Install tools

Reviewer: As mentioned earlier these are orthogonal to the Windows OS. So, we can keep the same list assuming the versions are portable across the different OSs.

So, feel free to add this: 
```
**Pre-requisites**
  * Maven Version 3.6.3
  * CMake 3.6 (or newer)
  * At least one of:
    * OpenCL: GPUs and CPUs >= 1.2, FPGAs >= 1.0
    * CUDA 9.0 +
  * GCC or clang/LLVM (GCC >= 5.5)
  * Python (>= 2.7)
  ```

### Maven
Download Apache Maven from the [official site](https://maven.apache.org/download.cgi) and extract it to the any location. Below it's assumed that Maven's home is `C:/Maven`.
### MSys2

> 1. Download [MSys2](https://www.msys2.org/) installer from the official site and run it. You may choose any installation directory, below it will be referred as `<MSYS2>`. IMPORTANT: the only executable you should use as a terminal is `<MSYS2>/mingw64.exe`, don't use others (msys2.exe, mingw32.exe) as this would lead to unpredictable results.

Reviewer 2: `Download the [MSys2](https://www.msys2.org/) installer  .. `

Remove this part:
`don't use others (msys2.exe, mingw32.exe) as this would lead to unpredictable results.`.

Author: ```suggestion
1. Download [MSys2](https://www.msys2.org/) installer from the official site and run it. You may choose any installation directory, below it will be referred as `<MSYS2>`. IMPORTANT: the only executable you should use as a terminal is `<MSYS2>/mingw64.exe`.
```

2. Update MSys2 _**system**_ packages. Start `<MSYS2>/mingw64.exe` and run the following command in the terminal:
```bash
pacman -Syu
```
You might need to execute it several times until you see that no updates found.

3. Update MSys2 _**default**_ packages. In the terminal window of `<MSYS2>/mingw64.exe` run:
```bash
pacman -Su
```
You might need to execute it several times until you see that no updates found.

4. Install necessary tools to MSys2. In the terminal window of `<MSYS2>/mingw64.exe` run:
```bash
pacman -S \
mingw-w64-x86_64-make           \
mingw-w64-x86_64-cmake          \

> mingw-w64-x86_64-gcc            \

Reviewer 2: Which GCC version is installed?

Author: Here is what shown on my pc:
```bash
$ cmake --version
cmake version 3.19.2
```
```bash
$ gcc --version
gcc.exe (Rev6, Built by MSYS2 project) 10.2.0
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

```bash
$ make --version
GNU Make 4.3
Built for x86_64-pc-msys
Copyright (C) 1988-2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

mingw-w64-x86_64-opencl-headers \
mingw-w64-x86_64-opencl-icd-git \
python make git
```

5. Create default Maven repository for MSys2 user:
```bash
cd ~
mkdir .m2
```

6. Create file `mvn` in `<MSYS2>/mingw64/bin` with any textual editor according to the template below:
```bash
#!/usr/bin/env bash
BASH_FILE=$(where bash)
SETTINGS=\"${BASH_FILE%\\*\\*\\*}${HOME}/.m2/settings.xml\"

C:/Maven/bin/mvn.cmd --settings ${SETTINGS} "$@"
```

> You should only change path to Maven executable to the actual location of Maven in your system.

Author: Such "proxy" is necessary - otherwise Maven will not see `settings.xml` - and requesting user to specify settings' file location is not an option here.

> You should only change path to Maven executable to the actual location of Maven in your system.

Reviewer: In my case, it works without the custom `mvn.cmd`. But I am ok to keep it in this PR.

## Get TornadoVM sources

> Clone the latest TornadoVM sources from Git repository using `<MSYS2>/mingw64.exe`:

Author: ```suggestion
Clone the latest TornadoVM sources from the GitHub [repository](https://github.com/beehive-lab/TornadoVM) using `<MSYS2>/mingw64.exe`:
```

> Clone TornadoVM 0.8 from Git repository using `<MSYS2>/mingw64.exe`:

Reviewer 2: Remove `0.8`. This should work for any version >= 0.8, shouldn't it?

Author: ```suggestion
Clone latest TornadoVM sources from Git repository using `<MSYS2>/mingw64.exe`:
```

```
cd D:/MyProjects
git clone https://github.com/beehive-lab/TornadoVM.git
cd TornadoVM
```
We will refer hereafter the directory with TornadoVM sources as `<TornadoVM>`.

## Get Graal 20.2.0

> Visist [GraalVM Community Edition 20.2.0](https://github.com/graalvm/graalvm-ce-builds/releases/tag/vm-20.2.0) and download either of the following builds:

Author: ```suggestion
Visit [GraalVM Community Edition 20.2.0](https://github.com/graalvm/graalvm-ce-builds/releases/tag/vm-20.2.0) and download either of the following builds:
```

- [graalvm-ce-java8-windows-amd64-20.2.0.zip](https://github.com/graalvm/graalvm-ce-builds/releases/download/vm-20.2.0/graalvm-ce-java8-windows-amd64-20.2.0.zip) -- Graal JDK 8 x64
- [graalvm-ce-java11-windows-amd64-20.2.0.zip](https://github.com/graalvm/graalvm-ce-builds/releases/download/vm-20.2.0/graalvm-ce-java11-windows-amd64-20.2.0.zip) -- Graal JDK 11 x64

Extract the downloaded file to any directory.

> ## Install NVIDIA CUDA and NVIDIA driver

Reviewer 2: I believe there are a few options. What TornadoVM needs is either
a) an OpenCL implementation (e.g. from CUDA, Intel OpenCL, AMD OpenCL)
b) a PTX implementation (CUDA) 

So it is possible to use, for example, Intel OpenCL instead of CUDA. I would reflect this in this document.

Author: It's reflected right after title

## Configuring build
Using any text editor create file `<TornadoVM>/etc/sources.env`  with the following content:
```bash
#!/bin/bash

# UPDATE PATH TO ACTUAL LOCATION OF THE JDK OR GRAAL (REMEMBER OF UNIX_STYLE SLASHES AND SPACES!!!)
export JAVA_HOME="C:/graalvm-ce-java8-20.2.0"  

## NEXT TWO LINES NECESSARY TO BUILD PTX (NVIDIA CUDA) BACKEND
## COMMENT THEM OUT OR JUST IGNORE IF YOU ARE NOT INTERESTED IN PTX BUILD
## OTHERWISE UPDATE 'CUDA_PATH' WITH ACTUAL VALUE (REMEMBER OF UNIX_STYLE SLASHES AND SPACES!!!)
export CUDA_PATH="C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.2"
export PTX_LDFLAGS=-L\"$CUDA_PATH/lib/x64\"

> # LEFT THE REST OF FILE 'AS IS'

Author: ```suggestion
# LEAVE THE REST OF FILE 'AS IS'
```

# DON'T ALTER!
export PATH=$PWD/bin/bin:$PATH               ## This directory will be automatically generated during Tornado compilation
export TORNADO_SDK=$PWD/bin/sdk              ## This directory will be automatically generated during Tornado compilation
CMAKE_FILE=$(where cmake | head -n 1)
export CMAKE_ROOT=${CMAKE_FILE%\\*\\*}
```
There are only 2 places you should adjust:
1. `JAVA_HOME` path that points to your Graal installation
2. `CUDA_PATH` pointing to your NVIDIA GPU Computing Toolkit (CUDA) - this one is necessary only for builds with PTX backend.

## Running build
Start `<MSYS2>/mingw64.exe` terminal, navigate to the `<TornadoVM>`  directory and execute build:
```bash
cd D:/MyProjects/TornadoVM
source etc/sources.env
make graal-jdk-8 BACKEND=ptx,opencl
```
The latest command, `make`, depends on the Graal version you are using:
1. For Graal JDK 8 use `make graal-jdk-8 BACKEND=ptx,opencl` (as above)
2. For Graal JDK 11 use `make graal-jdk-11 BACKEND=ptx,opencl` 

The `BACKEND` parameter has to be a comma-separated list of `ptx` and `opencl` options. You may build `ptx` only when NVIDIA GPU Computing Toolkit (CUDA) is installed.

## Verifying build

> Don't close `<MSYS2>/mingw64.exe` after build. Run the following command to see that TornadoVM is working:

Reviewer 2: after build -> after the build

Author: ```suggestion
Don't close `<MSYS2>/mingw64.exe` after the build. Run the following command to see that TornadoVM is working:
```

```bash
tornado --devices
```
You should see a list of OpenCL and/or CUDA devices available on your system.

> Now try to run a simple test. For Graal JDK 11:

Author: "Uses" or "using" here?

```bash 
tornado -m tornado.examples/uk.ac.manchester.tornado.examples.compute.MatrixMultiplication2D 512
```

> ... for Graal JDK 8:

Author: ```suggestion
To run examples with Graal JDK 8:
```

```bash 
tornado uk.ac.manchester.tornado.examples.compute.MatrixMultiplication2D 512
```

> If "sanity checks" from above work, you can execute complete test suite for TornadoVM:
> ```bash
> tornado-test.py -V
> ```

Author: ```suggestion
To run all unit-tests:
```bash
tornado-test.py -V
```
```

> If "sanity checks" from above work, you can execute complete test suite for TornadoVM:
> ```bash
> tornado-test.py -V
> ```

Author: ```suggestion
To run all unit-tests:
```bash
make tests
```

## Known issues
1. If you already have MSys2 installed and heavily customized you may experience issues with build or tests. We are suggesting to start with fresh MSys2 installation in this case and follow the instructions above. Most notably, make sure that you have no `mingw-w64-x86_64-python` installed - it prevents Python scripts that execute tests from running. Also, make sure that you have updated all GCC / Make / CMake packages mentioned.
2. If you see no output from `tornado --devices` this may be either of 2 reasons:
-- OpenCL / CUDA is misconfigured. Download any third-party tool for OpenCL / CUDA capabilities viewing and check that you can see your devices there. Sometimes order of installation of different OpenCL drivers matters - Intel OpenCL SDK may shadow NVIDIA OpenCL and alike.
-- You build native code of the library using wrong compiler, most probably you ran `<MSYS2>/msys2.exe` terminal instead of `<MSYS2>/mingw64.exe` . Please re-try with correct terminal (and therefore GCC) version.

> 3. If you see JVM crashes or `UnsatisfiedLinkError` or some `Error initializing DLL` during `tornado --devices` execution than it's definitely due to wrong GCC (and hence terminal) version used during build.

Reviewer 2: How we can install a specific version of GCC to avoid this problem?

Author: We will write above that GCC version should be checked to be 10 or higher

Reviewer 2: But TornadoVM can run with any version >-5.5. My question was about how can we download a specific version with the instructions you provide.

Author: We will merely reproduce MSys2 docs. For those, who install MSys2 for the first time all necessary versions (the latest and greatest) will be provided automatically. Those ones, who have existing installation are warned and may either update (and get the newest versions) or selectively update with specifying explicit version according to MSys2 docs (and we will point above which ones are reported to be valid).

You should only change path to Maven executable to the actual location of Maven in your system.

## Get TornadoVM sources

> Clone latest TornadoVM sources from Git repository using `<MSYS2>/mingw64.exe`:

Reviewer 2: latest -> the latest

Author: ```suggestion
Clone the latest TornadoVM sources from Git repository using `<MSYS2>/mingw64.exe`:
```
